### PR TITLE
Fix issue 259

### PR DESCRIPTION
**Pull request by akva2**
Two problems:

- non-empty interval for the random number generator for z starting coordinates
- problematic cc for the Deck class. workaround by moving to initializer list so move semantics are used.

### CODE REVIEW

File: opm/upscaling/CornerpointChopper.hpp [hunk 1]
```diff
@@ -40,12 +40,9 @@ namespace Opm
     class CornerPointChopper
     {
     public:
-        CornerPointChopper(const std::string& file)
+        CornerPointChopper(const std::string& file) :
+          deck_(parser_.parseFile(file, parseContext_))
```
Comment (bska) [inline opm/upscaling/CornerpointChopper.hpp]: Do we actually need a `parser_` and a `parseContext_` data member.  Can we not do
```C++
deck_(Parser{}.parseFile(file, ParseContext{}))
```
instead?

Moreover, this is an opportunity to also initialise the `metricUnits_` in the initialiser list.  I think we should do that too.

Comment (akva2) [reply opm/upscaling/CornerpointChopper.hpp]: i guess we do not need the members, i was just being safe. i'll check. and i'll move the units.

Comment (akva2) [reply opm/upscaling/CornerpointChopper.hpp]: updated.

**Comment by bska** [comment]
As for `Deck::Deck(const Deck& rhs)`, I think that's actually okay.  I do question using the defaulted/compiler generated `Deck::operator=(const Deck& rhs)` in a class with a user-defined copy constructor however.

**Comment by akva2** [comment]
ehm, yeah, i'm sorry. i meant assignment operator where i said cc.

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
also added a test. since any fixes to the assignment operator will be external, and author has been pinged, i will merge this now.

**Comment by joakim-hove** [comment]
I am quite offline today - superficially looking at this I did find the bug a bit surprising, but have nothing more to add now.

**Comment by atgeirr** [comment]
> I do question using the defaulted/compiler generated Deck::operator=(const Deck& rhs) in a class with a user-defined copy constructor however.

That missing operator is a dangerous bug, should be deleted I guess?

**Comment by bska** [comment]
>> I do question using the defaulted/compiler generated `Deck::operator=(const Deck& rhs)` in a class with a user-defined copy constructor however.
> 
> That missing operator is a dangerous bug, should be deleted I guess?

Probably.  I'll run a local build with the following patch
```diff
diff --git a/opm/parser/eclipse/Deck/Deck.hpp b/opm/parser/eclipse/Deck/Deck.hpp
index 43818cd2..09415371 100644
--- a/opm/parser/eclipse/Deck/Deck.hpp
+++ b/opm/parser/eclipse/Deck/Deck.hpp
@@ -131,6 +131,8 @@ namespace Opm {
 
             Deck( const Deck& );
 
+            Deck& operator=(const Deck& rhs) = delete;
+
             void addKeyword( DeckKeyword&& keyword );
             void addKeyword( const DeckKeyword& keyword );
```

**Comment by akva2** [comment]
it will work fine, already tested. just wasn't sure whether we wanted the operator or not.

**Comment by bska** [comment]
> I'll run a local build with [a disabled assignment operator]

> it will work fine, already tested. just wasn't sure whether we wanted the operator or not.

Sort of.  There are (at least) a couple of examples in opm-simulators that used the assignment operator.  I still haven't gone through all modules.

**Comment by akva2** [comment]
ehm, you will get one more similar error in upscaling. i forgot to build that repo.

**Comment by akva2** [comment]
yeah, i forgot to build simulators. sorry.

**Comment by atgeirr** [comment]
While at it, we should also delete the move constructor (currently private instead of deleted) and the move assignment operator. Either that, or implement them (no idea if that is difficult or not).

**Comment by andlaus** [comment]
> Either that, or implement them (no idea if that is difficult or not).

I agree. I'd argue for implementing them properly: as far as I see, this basically boils down to copy-and-pasting the copy constructor and adding `std::move()` to the right hand sides of assignments. Rinse and repeat for the constituent objects.

**Comment by bska** [comment]
> we should also delete the move constructor (currently private instead of deleted)

Class `Deck` does have a [private constructor](https://github.com/OPM/opm-common/blob/master/opm/parser/eclipse/Deck/Deck.hpp#L152) with the signature
```C++
Deck::Deck(std::vector<DeckKeyword>&&)
```
but no move constructor.  Since the class already has a user-defined copy constructor the move-aware operations (move constructor and move assignment) will not be generated by the compiler, meaning
```C++
std::move(deck)
```
will implicitly invoke the copy constructor.

**Comment by atgeirr** [comment]
> no move constructor

Thanks for pointing that out, I see that I misread that one.

**Comment by bska** [comment]
> I did find the bug a bit surprising,

In hindsight I think it's fairly clear.  The compiler-generated assignment operator will just do a blind copy  (well, assignment) of each individual data-member.  `DeckView::first` and `DeckView::last` are supposed to demarcate the beginning and end of a particular keyword range with the implicit assumption that the data is stored elsewhere.  Copying the iterator values in `Deck::operator=(const Deck&)` means that the range refers to the `Deck::keywordList` of the source object rather than the destination object.  That leads to dangling iterators (iterator invalidation) once the source object is destroyed at end of the assignment
```C++
deck_ = parser.parseFile(/* ... */)
```

**Pull request merged**